cmd/grafana-agent-service/service.go

```diff
@@ -0,0 +1,123 @@
+
+	minRestartBackoff time.Duration
+	maxRestartBackoff time.Duration
+}
+
+// serviceManagerConfig configures a service.
+type serviceManagerConfig struct {
+	// Path to the binary to run.
+	Path string
+
+	// Args of the binary to run, not including the command itself.
+	Args []string
+
+	// Dir specifies the working directory to run the binary from. If Dir is
+	// empty, the working directory of the current process is used.
+	Dir string
+
+	// Stdout and Stderr specify where the process' stdout and stderr will be
+	// connected.
+	//
+	// If Stdout or Stderr are nil, they will default to os.DevNull.
+	Stdout, Stderr io.Writer
+}
+
+// newServiceManager creates a new, unstarted serviceManager. Call
+// [service.Run] to start the serviceManager.
+//
+// Logs from the serviceManager will be sent to w. Logs from the managed
+// service will be written to cfg.Stdout and cfg.Stderr as appropriate.
+func newServiceManager(l log.Logger, cfg serviceManagerConfig) *serviceManager {
+	if l == nil {
+		l = log.NewNopLogger()
+	}
+
+	return &serviceManager{
+		log: l,
+		cfg: cfg,
+
+		minRestartBackoff: time.Second,
+		maxRestartBackoff: time.Minute,
+	}
+}
+
+// Run starts the serviceManager. The binary associated with the serviceManager
+// will be run until the provided context is canceled. The binary will be
+// restarted if it exits before the provided context is canceled.
+//
+// Intermediate restarts will increase with an exponential backoff, which
+// resets if the binary has been running for longer than the maximum
+// exponential backoff period.
+func (svc *serviceManager) Run(ctx context.Context) {
+	var (
+		restartBackoff = svc.minRestartBackoff
+		lastRestart    = time.Now()
+	)
+
+	for {
+		cmd := svc.buildCommand(ctx)
+
+		level.Info(svc.log).Log("msg", "starting program", "command", cmd.String())
+		err := cmd.Run()
+
+		// Handle the context being canceled before processing whether cmd.Run
+		// exited with an error.
+		if ctx.Err() != nil {
+			return
+		}
+
+		exitCode := cmd.ProcessState.ExitCode()
+
+		if err != nil {
+			level.Error(svc.log).Log("msg", "service exited with error; re-starting after wait period", "err", err, "exit_code", exitCode)
+		} else {
+			level.Info(svc.log).Log("msg", "service exited; re-starting after wait period", "exit_code", exitCode)
+		}
+
+		// Reset the backoff period if the process has been alive for at least
+		// twice the maximumRestartBackoff period.
+		now := time.Now()
+		if now.Sub(lastRestart) > 2*svc.maxRestartBackoff {
+			restartBackoff = svc.minRestartBackoff
+		}
+
+		level.Info(svc.log).Log("msg", "waiting before next restart", "restart_wait", restartBackoff)
+
+		// Wait for the backoff period.
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(restartBackoff):
+		}
+
+		// Increase the backoff period for the next iteration of the loop.
+		restartBackoff = restartBackoff * 2
+		if restartBackoff > svc.maxRestartBackoff {
+			restartBackoff = svc.maxRestartBackoff
```

Comment:
Does this mean it will try to restart forever? Maybe we have a maximum timeframe on it?

Reply:
Yeah, it will try to restart forever unless you stop the service.

Do you want me to add a maximum number of restarts?